Refactor jit::Operator

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36522 Better handling of alias analysis for aten::backward
* **#33905 Refactor jit::Operator**
* #33904 Fix either::map naming

jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

smessmer committed:
Refactor jit::Operator to more clearly distinguish the two possible states

jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -95,55 +112,73 @@ struct TORCH_API Operator {
-    }
-    AT_ASSERT(node != nullptr);
-    return op_creator_(node);
+    return op_.fold<Operation>([] (const C10Operator& op) {
```
I still find this lambda expressions hard to read. This code is not perf-critical (as it happens in initialization only) - why not just write an if statement? :)

smessmer replied:
I think this is on the hot path for the jit compiler (i.e. relevant for compilation time).

I read those similar to how I read pattern matching statements in more descriptive languages, that makes it ok to read.

dzhulgakov reviewed (changes requested):
Sure, but there are broken tests

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

bhosmer commented:
I haven't had a chance to review this, will do tomorrow, but first a high-level question: how much longer do we expect JIT-only operators to exist?

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

ezyang commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -40,51 +40,60 @@ using OperationCreator = Operation (*)(const Node*);
+    c10::OperatorHandle handle_;
+    Operation op_;
+  };
+  struct UnmaterializedFunctionSchema final {
```
unmaterialized seems to be a new thing in this PR, so it would be helpful to briefly comment what this is. (It's a just a schema that hasn't been parsed yet, righT? Maybe UnparsedFunctionSchema is more direct.)

ezyang commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -40,51 +40,60 @@ using OperationCreator = Operation (*)(const Node*);
+    mutable c10::optional<c10::AliasAnalysisKind> alias_analysis_;
+  };
+  struct JitOnlyOperator final {
+    mutable c10::either<FunctionSchema, UnmaterializedFunctionSchema> schema_;
```
Probably worth mentioning that you can only transition from UnmaterializedFUnctionSchema to FunctionSchema, but not vice versa?

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]

bhosmer reviewed (commented):
Thinking there might be a simpler way to do the data modeling to avoid nested eithers, but that's for later. :)

smessmer commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -91,86 +103,89 @@ struct TORCH_API Operator {
+    const FunctionSchema& schemaRef = schema();
+    c10::AliasAnalysisKind alias_analysis = schemaRef.aliasAnalysis();
+
+    if(C10_LIKELY(schemaRef.name() != "aten::backward")) {
```
separated out into https://github.com/pytorch/pytorch/pull/36522

smessmer committed:
Update on "Refactor jit::Operator"


jit::Operator is semantically either a c10 op or a jit-only op but that is represented in a set of member variables with intricate invariants about their values.
Making this explicitly represented in a c10::either reduces the number of possible states, removing many of the invalid ones.

Similarly, if it is a jit-only op, there were schema_string_ and schema_ of which only one could be set at any time. Using a c10::either there too.

Differential Revision: [D20147487](https://our.internmc.facebook.com/intern/diff/D20147487/)

[ghstack-poisoned]
